internal/cosesign1/check.go

```diff
@@ -0,0 +1,282 @@
+	var chain []*x509.Certificate
+	var chainIA []interface{}
+	if isDERChain(chainDER) {
+		chainIA = chainDER.([]interface{})
+	} else {
+		chainIA = append(chainIA, chainDER)
+	}
+
+	for _, element := range chainIA {
+		cert, err := x509.ParseCertificate(element.([]byte))
+		chain = append(chain, cert)
+		if err != nil {
+			if verbose {
+				log.Print("Parse certificate failed: " + err.Error())
+			}
+			return nil, err
+		}
+	}
+
+	// A reasonable chain will have a handful of certs, typically 3 or 4,
+	// so limit to an arbitary 100 which would be quite unreasonable
+	chainLen := len(chain)
+	if chainLen > 100 || chainLen < 1 {
+		return nil, fmt.Errorf("unreasonable number of certs (%d) in COSE_Sign1 document", chainLen)
+	}
+
+	// We need to split the certs into root, leaf and intermediate to use x509.Certificate.Verify(opts) below
+	rootCerts := x509.NewCertPool()
+	intermediateCerts := x509.NewCertPool()
+	var leafCert *x509.Certificate
+	var rootCert *x509.Certificate
+
+	if verbose {
+		log.Print("Certificate chain:")
+	}
+	// since the certs come from the ordered HeaderLabelX5Chain we can assume chain[0] is the leaf,
+	// chain[len-1] is the root, and the rest are intermediates.
+	for i, cert := range chain {
+		if i == 0 {
+			leafCert = cert
+			if verbose {
+				log.Println(x509ToPEM(cert))
+			}
+		} else if i == chainLen-1 {
+			rootCert = cert
+			rootCerts.AddCert(rootCert)
+			if verbose {
+				log.Println(x509ToPEM(cert))
+			}
+		} else {
+			intermediateCerts.AddCert(cert)
+			if verbose {
+				log.Println(x509ToPEM(cert))
+			}
+		}
+	}
+
+	opts := x509.VerifyOptions{
+		Intermediates: intermediateCerts,
+		Roots:         rootCerts,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny}, // To be removed when I have a decent set of examples.
+		// consider CurrentTime time.Time set to be the time the cose message was signed so we are checking the certs were valid then rather than now. Maybe not TBD
+	}
+
+	/*
+		Until we have some production certs allow the certificate check to fail
+	*/
+
+	_, err = leafCert.Verify(opts)
+
+	if err != nil {
+		return nil, fmt.Errorf("certificate chain verification failed")
+	}
+
+	var leafCertBase64 = x509ToBase64(leafCert) // blob of the leaf x509 cert reformatted into pem (base64) style as per the fragment policy rules expect
+	var leafPubKey = leafCert.PublicKey
+	var leafPubKeyBase64 = keyToBase64(leafPubKey)
+
+	var chainPEM string
+	for i, c := range chain {
+		if i > 0 {
+			chainPEM += "\n"
+		}
+		chainPEM += x509ToPEM(c)
+	}
+
+	var results = UnpackedCoseSign1{
+		Pubcert:     leafCertBase64,
+		Feed:        feed,
+		Issuer:      issuer,
+		Pubkey:      leafPubKeyBase64,
+		ChainPem:    chainPEM,
+		ContentType: msg.Headers.Protected[cose.HeaderLabelContentType].(string),
+		Payload:     msg.Payload,
+		CertChain:   chain,
+	}
+
+	if err != nil {
```

Comment:
This error is already returned after the verify operation above.

Reply:
ok